pacsman/filesystem_dev_client.py

```diff
@@ -48,78 +75,162 @@ def __init__(self, dicom_dir: str, dicom_source_dir: str, *args, **kwargs) -> No
         os.makedirs(self.dicom_dir, exist_ok=True)
         self.dicom_source_dir = dicom_source_dir
 
-        self.dicom_datasets: Dict[str, Dataset] = {}
+        self.cached_dicom_datasets: Dict[str, Dataset] = {}
 
-        for dicom_file in glob.glob(f'{dicom_source_dir}/**/*.dcm', recursive=True):
-            self._read_and_add_data_set(dicom_file)
+        # load and use the index if it is present and its hash matches the current dir
+        index_path = self._filepath(INDEX_FILENAME)
+        if os.path.exists(index_path):
+            with open(index_path, 'rb') as f:
+                self.index = pickle.load(f)
+                if self.index.dicom_source_dir_hash != self._dicom_source_dir_hash():
```

Comment:
Just wondering, but why did you decide to put the hash in the index file, vs putting it in the file name?

E.g., is it so that we don't accumulate index files?

If the hash was in the filename, we could check it without loading the file.

Reply:
Having one index file name makes it easier to keep track, put in things like `.gitignore`, etc. The index file is pretty small (less than 1MB for 150+ series), so loading the file to check the hash is not a big slowdown.